### ISSUE TITLE
Create more unit tests

### ISSUE THREAD

**Issue by milescalabresi** [learning-unlimited/ESP-Website#3780]
Go through one of our test/deployment documents and create a ticket for each item that isn't currently tested in our automatic pipeline

**Comment by Diya910** [comment]
Sure, test cases are the subtle but very core component in preventing code clashes. I would love to deep dive in the code and write some necessary test cases. I will proceed with the ones that are being tagged first. Can I proceed?

**Comment by willgearty** [comment]
Hi @Diya910, yes please good ahead and make PRs for small subsets of tests. The ones that are already tagged would be a great place to start.

**Comment by Diya910** [comment]
Hey @willgearty , I've been looking into #3977 and wanted to share my understanding of the test cases needed before I start writing them — please correct me if I'm off anywhere!

Here's what I'm planning to test:

**Recipient Resolution**
- learningu.org alias resolves to the correct recipients
- PlainRedirect lookup returns the correct destination
- ESPUser lookup finds the right user by email
- Empty recipient list after filtering is handled gracefully
- Null redirect destinations are handled (no crash)

**Sender Validation**
- Email from a valid site user gets forwarded successfully
- Email from an unknown sender (no account on site) gets rejected and a bounce is sent
- Malformed or unparseable From address is handled gracefully
- Sender with an email not associated with any learningu.org account gets rejected

**Multi-Account Sender Logic**
- When a sender has multiple accounts, the correct one is picked

**Attachment Handling**
- Email with attachments gets correctly extracted and forwarded
- An attachment that fails extraction gets skipped with logging and doesn't crash

**Edge Cases**
- Missing From header is handled gracefully
- Recipients list is empty after all filtering steps

A few questions before I proceed:
1. Should I mock SendGrid calls throughout, or is there a test helper already set up for that?
2. Where should the test file live — alongside existing mailgate files or somewhere else in the test suite?

Let me know if I'm missing anything or misunderstood any part of the logic.

**Comment by Oval17** [comment]
Hey @Diya910 
Just to add my two cents to the discussion, you can view the PR that was raised by me..
Also as mostly those test cases would be written by AI, kindly make sure that when we do prompting we should mention that it should cover the codes(important functions) and should not write always truthy test cases as AI hallucinates a lot in these cases.
Please make sure that coverage is overall high.
Thanks

**Comment by willgearty** [comment]
@Diya910 thanks for reaching out and for your interest in contributing to our project! That issue is already assigned to someone else (and is currently waiting on a PR to be merged). However, this is all really great and would be useful to have over on that sub-issue.

**Comment by Srinidhi188** [comment]
Hi @willgearty @milescalabresi  , I’ve finished setting up the ESP Website locally, including legacy dependencies and PostgreSQL, and I’m able to run the full test suite successfully.

I noticed several regression testing issues are already assigned. I’d still love to contribute — especially by identifying untested code paths or adding additional unit tests where coverage is low.

Please let me know if there are specific modules that would benefit from extra test coverage, or if I should open a new issue for any gaps I find.

Thanks!

**Comment by willgearty** [comment]
@Srinidhi188 thanks for your interested in helping us improving our tests! I would suggest looking at our code coverage report and identifying areas that need more tests: https://app.codecov.io/gh/learning-unlimited/ESP-Website.

**Comment by pushkarkumar06** [comment]
Hi @willgearty ! 👋
I’d love to work on this issue and help improve the test coverage for the project. If it’s okay, please assign this issue #3780 to me. I’m ready to start and follow the project’s guidelines.
Thanks!

**Comment by willgearty** [comment]
@pushkarkumar06 this is a large parent issue (that is actually listed as a GSoC project). You're welcome to work on adding tests for an individual module or part of our codebase, but we won't be assigning this parent issue to anyone.

**Comment by yuvraj-kolkar17** [comment]
Hi @willgearty ,
I’ve explored the project and reviewed the code coverage report. I identified a few modules where the coverage is relatively low, and I’d really love to contribute by improving the test coverage in those areas.
I also have prior **experience contributing to CNCF projects**, where I worked on improving test coverage and related issues, so I’m comfortable following the contribution guidelines and maintaining quality standards.

I’m a bit confused about the process should I open a separate issue for the specific module with low coverage, or can I directly start writing test cases and submit a PR?
Your clarification would really help.
Thank you!

**Comment by moksha-hub** [comment]
@willgearty I'd like to add tests for classchange.py and studentregsanity.py both at 0% coverage. May I create a sub-issue for this? 
And also like to work on them if assigned and accepted.
Thank you.

**Comment by Oval17** [comment]
Hi @yuvraj-kolkar17 @moksha-hub ,
Yes you can proceed with creating and writing tests if you see its coverage is 0 , it will be reviewed by maintainers as they get time.
Please make sure that our aim is not to just increase the coverage percentage but to have quality test cases which actually prove to be sturdy if new codes are added.

**Comment by SKM2227229725** [comment]
Hi @willgearty,
I’ve been exploring the website and codebase and have identified a few UI areas that could be improved.
I’m currently focusing on resolving specific existing issues and will be raising a PR soon.
I plan to continue with smaller improvements before drafting a full proposal.

Best,
Shailesh Kumar

**Comment by Dhirenderchoudhary** [comment]
@willgearty can i do some of fixes from this issue ?
can i get it assigned

**Comment by willgearty** [comment]
@Dhirenderchoudhary this parent issue is a GSoC project, but you're welcome to make an issue for a particular module or file that needs tests and ask to be assigned to that.

**Comment by ISHAN9876** [comment]
Hi @willgearty

I noticed that the admintoolsmenu.py and admintoolsdash.py currently have 0% coverage and I'd love to help improve that.
Could you please assign me

**Comment by Dhirenderchoudhary** [comment]
@ISHAN9876 sorry but I am currently doing that part and raise PR today 
Sorry for the inconvenience.

**Comment by ISHAN9876** [comment]
How about mailman , dbmail, random/views.py and customforms? can i be assigned those? @willgearty

**Comment by willgearty** [comment]
@ISHAN9876 please make a separate issue for those files and I can assign you. Thanks.

**Comment by priyasnhibisht** [comment]
Hi @willgearty ! 
I was checking the Codecov coverage and noticed that dbmail_cron.py has no tests.
I'd like to add unit tests for this file.
Should I create a seperate issue for this.
Thank you.

**Comment by ISHAN9876** [comment]
@priyasnhibisht #4481 i have already been assigned and i am working on it

**Comment by priyasnhibisht** [comment]
@ISHAN9876 Thanks for telling me.
I'll check out some other files.

**Comment by Ashutosh9470** [comment]
hey @willgearty 
I’d like to contribute to expanding regression test coverage.
Is there a specific untested module or sub-issue that is currently unassigned that I can start with?

**Comment by willgearty** [comment]
I'm not sure. You can take a look at our [coverage stats](https://app.codecov.io/gh/learning-unlimited/ESP-Website), pick a file that has no coverage, then look to see if there is already an issue for it.

**Comment by Ashutosh9470** [comment]
Hey @willgearty 
Thanks for the suggestion! I ran coverage locally and identified a few low-coverage utility files. I’m starting by adding tests for esp.utils.sanitize to improve coverage there. I’ll open a PR shortly.

**Comment by Sugandh04** [comment]
I noticed that esp/admintoolsmenu.py currently has 0% coverage in the Codecov report.
I've opened an issue to add unit tests for it and would like to work on improving coverage for that module.

**Comment by ISHAN9876** [comment]
@sugandh a week earlier dhirendra told me he was on this issue or else i would have already submitted a pr 😐, Good for you though

**Comment by Dhirenderchoudhary** [comment]
@Sugandh04 I am on that issue and Pr has been opened so please don't work on that.

**Comment by Dhirenderchoudhary** [comment]
> @sugandh a week earlier dhirendra told me he was on this issue or else i would have already submitted a pr 😐, Good for you though

Yes yes 🙌 don't worry I'm working on that.

**Comment by Sugandh04** [comment]
I noticed that esp/dbmail_cron.py currently has 0% coverage in the Codecov report.
I've opened an issue to add unit tests for it and would like to work on improving coverage for that module.

**Comment by Dhirenderchoudhary** [comment]
> I noticed that esp/dbmail_cron.py currently has 0% coverage in the Codecov report. I've opened an issue to add unit tests for it and would like to work on improving coverage for that module.

#4481  issued is already raised and @ISHAN9876 is working on them from past days

**Comment by akashthakrele** [comment]
Hi @milescalabresi  I'm applying for GSoC 2026 and the Regression Tests Expansion project caught my eye. I'm fairly new to open source but comfortable with Python and keen to get started. Would it be okay if I picked up one of the sub-issues? Happy to take any guidance on where to begin.

**Comment by akashthakrele** [comment]
@willgearty @milescalabresi — I noticed all the sub-issues are currently assigned. Are there other untested modules you'd suggest a beginner start with? Happy to take direction on where to contribute.

**Comment by willgearty** [comment]
@akashthakrele there are still plenty of other issues to work on: https://github.com/learning-unlimited/ESP-Website/issues?q=is%3Aissue%20state%3Aopen%20no%3Aassignee

**Comment by akashthakrele** [comment]
@willgearty  Thanks! I've gone through the list and I'm setting up the project locally now. I'll find an appropriate issue and get assigned before submitting any PR.

**Comment by Srinidhi188** [comment]
Hi, I’ve been working on this issue through PR #4333. Could I please be assigned to issue #3780? Thanks!

**Comment by willgearty** [comment]
@Srinidhi188 you'll need to make a separate issue for this. You can then assign yourself to it using the /assign command.

**Comment by Abha377** [comment]
Hi, I would like to work on this issue as part of my GSoC 2026 preparation.

I have started exploring the codebase and existing test structure, and I plan to identify untested areas and contribute by adding unit tests to improve coverage.

Please let me know if there are any specific modules or priorities I should focus on first. I would really appreciate any guidance.

Thank you!

**Comment by Abha377** [comment]
> ⚠️ Sorry @Abha377, this issue is labeled **gsoc 2026** and is not currently available for assignment.
> 
> 📖 [Getting Started](https://github.com/learning-unlimited/ESP-Website/wiki#getting-started) · [Contribution Rules](https://github.com/learning-unlimited/ESP-Website/wiki#rules) · [Code of Conduct](https://github.com/learning-unlimited/ESP-Website/blob/main/CODE_OF_CONDUCT.md)

Thank you for the clarification!

I will go ahead and start exploring the codebase and contribute by adding tests to improve coverage. I’ll submit a PR once I have something ready.

Please feel free to suggest any areas that might benefit from additional test coverage.

Thanks again!

**Comment by Oval17** [comment]
Hi @CMSteam-hm @HiddendarkViper07 
You guys can pick individual issues and create tests, or see if theres code that doesnt have test coverage.

**Comment by ishandabi854-pixel** [comment]
# PR Draft for Issue #3780: Create more unit tests

## Overview
This PR significantly improves the test coverage of the ESP Website by introducing comprehensive unit tests for several critical but previously untested modules.

## Related Issues
- #3780: Create more unit tests
- #3977: Tests for email relay
- #5243: Add unit tests for BigBoardModule and OnSiteAttendance
- #5288: Add unit tests for studentregsanity.py

## Why Unit Testing is used?
Unit testing is a fundamental practice in modern software development for several reasons:
1.  **Bug Prevention**: Tests catch regressions and bugs early in the development cycle, before they reach production.
2.  **Code Quality**: Writing tests often forces developers to write cleaner, more modular code that is easier to maintain.
3.  **Documentation**: Tests serve as living documentation, demonstrating how specific modules and functions are expected to behave.
4.  **Confidence in Refactoring**: With a robust test suite, developers can refactor legacy code with confidence, knowing that any breaking changes will be immediately flagged.

## Implementation Details

### 1. Student Registration Sanity (`test_studentregsanity.py`)
- Tests the `StudentRegSanityController` which handles cleaning up invalid registrations (e.g., walk-ins and lunch blocks).
- Verifies both report generation (`fake=True`) and actual data removal (`fake=False`).

### 2. Big Board Module (`test_bigboard.py`)
- Tests the `BigBoardModule` which provides aggregate statistics for the program dashboard.
- Verifies student enrollment counts, check-in status tracking, and overall registration metrics.

### 3. Email Relay Logic (`tests_relay.py`)
- Tests the `UserEmail` relay handler used by the mailgate system.
- Verifies that emails sent to system addresses are correctly processed and routed to the intended recipients.

---

## Files Included in this PR
- `esp/esp/program/tests/test_studentregsanity.py`: New tests for registration sanity.
- `esp/esp/program/tests/test_bigboard.py`: New tests for the program dashboard.
- `esp/esp/dbmail/tests_relay.py`: New tests for email relay functionality.

**Note**: To integrate these tests, move the files from this draft folder to their respective directories in the `ESP-Website` codebase and run `python manage.py test`.

**Comment by willgearty** [comment]
@ishandabi854-pixel This issue is part of a GSoC project. Please submit a GSoC project proposal if you would like to work on this issue. If you keep posting spam, I will ban you from this repository.

**Comment by HiddendarkViper07** [comment]
@willgearty Hi, I am Vaibhav, a student and developer skilled in Python, Django, and Java. I am interested in contributing to your project for GSoC.

**Comment by ishandabi854-pixel** [comment]
I want to contribute on this project what can I do for that

**Comment by mohanaddalol** [comment]
Hi  @willgearty @milescalabresi,

I am a final-year Computer Science (Cybersecurity) student, and I have submitted a GSoC proposal for the Regression Tests Expansion project. I have been reviewing the ESP Website codebase and exploring the current testing structure as part of my preparation.

I am particularly interested in improving test coverage and understanding how different modules are validated, and I am ready to start contributing. Please let me know if there is a good issue I can begin with.

Thank you!

**Comment by willgearty** [comment]
@mohanaddalol More than 100 available issues right here: https://github.com/learning-unlimited/ESP-Website/issues?q=is%3Aissue%20state%3Aopen%20no%3Aassignee%20-label%3A%22GSoC%202026%22%20-label%3A%22Idea%20for%20Discussion%22%20-label%3A%22Server%20setup%22%20-label%3AFormstack%20-label%3A%22Email%20(mailman)%22&page=1.

**Comment by KavinRajNV** [comment]
Hi @mcalabr2 @willgearty — I'm Kavin Raj N V, a second-year AI & Data Science student interested in applying for the Regression Tests Expansion project for GSoC 2026.
I've been contributing to the codebase recently (PRs #5533 and #5542) to get familiar with the project structure. I've read through the relevant issues (#3773, #599, #794) and the existing test files.
I'd love to get your guidance on what you'd expect from a strong proposal and where you'd suggest focusing test coverage first. Is there anything specific you'd like to see in the application?
Thank you!

**Comment by Piyushxgit** [comment]
Hi! I am applying for GSoC for this project and have put together a PR addressing this. I would love to be assigned to this issue so I can own seeing it through review!

**Comment by NaniToka** [comment]
Hi! I am Toka Nani, a GSoC 2026 applicant. I have submitted a proposal for the Regression Tests Expansion project. I would like to be assigned to this issue so I can contribute. I have already made a pull request (#5570). Please assign me!
